### PR TITLE
[SPARK-23599][SQL] Add a UUID generator from Pseudo-Random Numbers

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

This patch adds a UUID generator from Pseudo-Random Numbers. We can use it later to have deterministic `UUID()` expression.

## How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @hvanhovell

**Comment by kiszk** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@kiszk The failed test seems from #20779. Is it still flaky?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * For the algorithm, see RFC 4122: A Universally Unique IDentifier (UUID) URN Namespace,
+ * section 4.4 "Algorithms for Creating a UUID from Truly Random or Pseudo-Random Numbers".
+ */
+case class RandomUUIDGenerator(random: Random) {
+  def getNextUUID(): UUID = {
+    val mostSigBits = (random.nextLong() & 0xFFFFFFFFFFFF0FFFL) | 0x0000000000004000L
+    val leastSigBits = (random.nextLong() | 0x8000000000000000L) & 0xBFFFFFFFFFFFFFFFL
+
+    new UUID(mostSigBits, leastSigBits)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala]: I think we need to use a different RNG. `java.util.Random` only has 48 bits of state, which is less than the 122 bits we need for random number generation. Something like [PCG](http://pcg-random.org/) or a Mersenne twister would work.

**Comment by kiszk** [comment]
@viirya umm, it may eat more memory sometime ...

**Comment by hvanhovell** [comment]
@kiszk is it taking more memory because of the test? If it does can we make the test case smaller?

**Comment by kiszk** [comment]
I think so. Unfortunately, when I changed the size of test, the problem cannot be reproduced in my environment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala]: Ok. Mersenne Twister is used in the update.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * This class is used to generate a UUID from Pseudo-Random Numbers.
+ *
+ * For the algorithm, see RFC 4122: A Universally Unique IDentifier (UUID) URN Namespace,
+ * section 4.4 "Algorithms for Creating a UUID from Truly Random or Pseudo-Random Numbers".
+ */
+case class RandomUUIDGenerator(randomSeed: Long) {
+  private val random = new MersenneTwister(randomSeed)
+
+  def getNextUUID(): UUID = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala]: Perhaps we should also create a version that creates a  UTF8String directly.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RandomUUIDGenerator.scala]: Sounds good. I've added it.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
ping @hvanhovell Is there any more comments? Thanks.

**Review by hvanhovell — APPROVED** [review]
LGTM - merging to master. Thanks!

**Comment by viirya** [comment]
@hvanhovell Thanks for merging this! I will continue to work on make use of this UUID generator in UUID expression.

**Pull request closed without merging**